examples/single_offer/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+#![no_std]
+
+#[cfg(feature = "external")]
+extern crate std;
+
+pub mod external;
+mod test;
+use stellar_contract_sdk::{
+    contractimpl, contracttype, BigInt, Binary, Env, EnvVal, IntoEnvVal, RawVal, Symbol, Vec,
+};
+use stellar_token_contract::public_types::{Identifier, KeyedAuthorization, U256};
+
+#[derive(Clone, Copy)]
+#[repr(u32)]
+pub enum DataKey {
+    SellToken = 0,
+    BuyToken = 1,
+    Admin = 2,
+    Price = 3,
+}
+
+impl IntoEnvVal<Env, RawVal> for DataKey {
+    fn into_env_val(self, env: &Env) -> EnvVal {
+        (self as u32).into_env_val(env)
+    }
+}
+
+// Price is 1 unit of selling in terms of buying. For example, if you wanted
+// to sell 30 XLM and buy 5 BTC, the price would be Price{n: 5, d: 30}.
+#[derive(Clone)]
+#[contracttype]
+struct Price {
+    pub n: u32,
+    pub d: u32,
+}
+
+fn get_contract_id(e: &Env) -> Identifier {
+    Identifier::Contract(e.get_current_contract().into())
+}
+
+fn get_sell_token(e: &Env) -> Binary {
+    e.get_contract_data(DataKey::SellToken)
+}
+
+fn get_buy_token(e: &Env) -> Binary {
+    e.get_contract_data(DataKey::BuyToken)
+}
+
+fn get_balance(e: &Env, contract_id: Binary) -> BigInt {
+    let mut args: Vec<EnvVal> = Vec::new(&e);
+    args.push(get_contract_id(e).into_env_val(&e));
+    e.call(contract_id, Symbol::from_str("balance"), args)
+}
+
+fn get_balance_sell(e: &Env) -> BigInt {
+    get_balance(&e, get_sell_token(&e))
+}
+
+fn get_balance_buy(e: &Env) -> BigInt {
+    get_balance(&e, get_buy_token(&e))
+}
+
+fn put_sell_token(e: &Env, contract_id: U256) {
+    e.put_contract_data(DataKey::SellToken, contract_id);
+}
+
+fn put_buy_token(e: &Env, contract_id: U256) {
+    e.put_contract_data(DataKey::BuyToken, contract_id);
+}
+
+fn put_price(e: &Env, price: Price) {
+    e.put_contract_data(DataKey::Price, price);
+}
+
+fn get_price(e: &Env) -> Price {
+    e.get_contract_data(DataKey::Price)
+}
+
+fn transfer(e: &Env, contract_id: Binary, to: Identifier, amount: BigInt) {
+    let mut args = Vec::new(e);
+    args.push(KeyedAuthorization::Contract.into_env_val(e));
+    args.push(to.into_env_val(e));
+    args.push(amount.into_env_val(e));
+    e.call::<()>(contract_id, Symbol::from_str("xfer"), args);
```

Comment:
Are the three args able to be used in a tuple? I think you should be able to use tuple syntax with a `.into()` or `.try_into().unwrap()`. This works too, just wondering if that shorthand isn't working.

There is also the `vec![&e, ...]` you can use too.

Reply:
`vec!` requires `std`. The tuple idea doesn't work either. Is that expected to work with sdk `Vec`?

Reply:
The std implementation of `vec!` builds a std Vec. The SDK has an implementation of `vec!` that builds an SDK `Vec`. It looks like you're using the SDK `Vec` here, so it should work.

Reply:
Ah yeah I missed that. `stellar_contract_sdk::vec![&e, KeyedAuthorization::Contract.into_env_val(e), to.into_env_val(e), amount.into_env_val(e)]` works. I'd prefer to keep it the way it is in this PR since this is also done in a few other places and I don't want this PR to get stale due to the move, but I'm fine with changing this is another PR.